src/main/java/org/cryptomator/linux/tray/AppindicatorTrayMenuController.java

```diff
@@ -0,0 +1,123 @@
+package org.cryptomator.linux.tray;
+
+import org.cryptomator.integrations.common.CheckAvailability;
+import org.cryptomator.integrations.common.OperatingSystem;
+import org.cryptomator.integrations.common.Priority;
+import org.cryptomator.integrations.tray.ActionItem;
+import org.cryptomator.integrations.tray.SeparatorItem;
+import org.cryptomator.integrations.tray.SubMenuItem;
+import org.cryptomator.integrations.tray.TrayIconLoader;
+import org.cryptomator.integrations.tray.TrayMenuController;
+import org.cryptomator.integrations.tray.TrayMenuException;
+import org.cryptomator.integrations.tray.TrayMenuItem;
+import org.purejava.appindicator.MemoryAllocator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.lang.foreign.MemorySegment;
+import java.lang.foreign.SegmentScope;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Paths;
+import java.util.List;
+import java.util.function.Consumer;
+
+import static org.purejava.appindicator.app_indicator_h.*;
+
+@Priority(1000)
+@OperatingSystem(OperatingSystem.Value.LINUX)
+public class AppindicatorTrayMenuController implements TrayMenuController {
+
+	private static final Logger LOG = LoggerFactory.getLogger(AppindicatorTrayMenuController.class);
+
+	private static final SegmentScope SCOPE = SegmentScope.global();
+	private MemorySegment indicator;
+	private MemorySegment menu = gtk_menu_new();
+
+	@CheckAvailability
+	public static boolean isAvailable() {
+		return MemoryAllocator.isLoadedNativeLib();
+	}
+
+	@Override
+	public void showTrayIcon(URI uri, Runnable runnable, String s) throws TrayMenuException {
+		indicator = app_indicator_new(MemoryAllocator.ALLOCATE_FOR("org.cryptomator.Cryptomator"),
+				MemoryAllocator.ALLOCATE_FOR(getAbsolutePath(getPathString(uri))),
+				APP_INDICATOR_CATEGORY_APPLICATION_STATUS());
+		gtk_widget_show_all(menu);
+		app_indicator_set_menu(indicator, menu);
+		app_indicator_set_status(indicator, APP_INDICATOR_STATUS_ACTIVE());
+	}
+
+	@Override
+	public void updateTrayIcon(Consumer<TrayIconLoader> iconLoader) {
+		TrayIconLoader.FreedesktopIconName callback = this::updateTrayIconCallback;
+		iconLoader.load(callback);
+	}
+
+	private void updateTrayIconCallback(String s) {
+		app_indicator_set_icon(indicator, MemoryAllocator.ALLOCATE_FOR(s));
```

Comment:
This is a good example, where we want the lifetime of the c string to be >= the duration of the method invocation. What we can use here is a confined arena:

```suggestion
		try (var arena = MemoryArena.openConfined()) {
			app_indicator_set_icon(indicator, arena.allocateUtf8String(s));
		}
```

I have the gut feeling that what `MemoryAllocator.ALLOCATE_FOR` is doing there is not as it is supposed to work with `RuntimeHelper` not being a public API.

Reply:
In fact, _all_ the strings in this class are merely passed as parameters to some native function. They all can be allocated by a short-lived, confined scope.

Reply:
Another alternative which we used in jextract, was to use an ad-hoc allocator for one-off strings that would just slice off a given segment:

https://github.com/openjdk/jextract/blob/master/src/main/java/org/openjdk/jextract/clang/LibClang.java#L89

(e.g. just have a "constant" segment, and "allocate" by slicing a piece off it - assuming the uses are well-behaved - e.g. the native code doesn't want to hold onto the string bits - it should work).

Reply:
@mcimadamore Thanks for the hint, didn't know this option before. Just read the JavaDoc for the factories defined in SegmentAllocator and believe this is a good option as well.

Performance isn't critical in this application, though. We need a dozen strings at app start, that's it.

Reply:
Thanks @overheadhunter and @mcimadamore for the hints.

It's implemented.